tst/com/amazon/corretto/crypto/provider/test/CounterKdfTest.java

```diff
@@ -0,0 +1,120 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.corretto.crypto.provider.test;
+
+import static com.amazon.corretto.crypto.provider.test.TestUtil.bcDigest;
+import static com.amazon.corretto.crypto.provider.test.TestUtil.getEntriesFromFile;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assumptions.assumeFalse;
+
+import com.amazon.corretto.crypto.provider.CounterKdfSpec;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.stream.Stream;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import org.bouncycastle.crypto.generators.KDFCounterBytesGenerator;
+import org.bouncycastle.crypto.macs.HMac;
+import org.bouncycastle.crypto.params.KDFCounterParameters;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.parallel.Execution;
+import org.junit.jupiter.api.parallel.ExecutionMode;
+import org.junit.jupiter.api.parallel.ResourceAccessMode;
+import org.junit.jupiter.api.parallel.ResourceLock;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+@ExtendWith(TestResultLogger.class)
+@Execution(ExecutionMode.CONCURRENT)
+@ResourceLock(value = TestUtil.RESOURCE_GLOBAL, mode = ResourceAccessMode.READ)
+public class CounterKdfTest {
+  @Test
+  public void concatenationKdfsAreNotAvailableInFipsMode() {
+    Stream.of("CounterKdfWithHmacSHA256", "CounterKdfWithHmacSHA384", "CounterKdfWithHmacSHA512")
+        .forEach(
+            alg -> {
+              try {
+                assertNotNull(SecretKeyFactory.getInstance(alg, TestUtil.NATIVE_PROVIDER));
+              } catch (final NoSuchAlgorithmException e) {
+                assertTrue(TestUtil.isFips());
+              }
+            });
+  }
+
+  @Test
+  public void secretLengthCannotBeZero() {
+    assertThrows(IllegalArgumentException.class, () -> new CounterKdfSpec(new byte[0], 1, "name"));
+  }
+
+  @Test
+  public void outputLengthCannotBeZeroOrNegative() {
+    assertThrows(IllegalArgumentException.class, () -> new CounterKdfSpec(new byte[1], 0, "name"));
+    assertThrows(IllegalArgumentException.class, () -> new CounterKdfSpec(new byte[1], -1, "name"));
+  }
+
+  // The rest of the tests are only available in non-FIPS mode.
+  @Test
+  public void concatenationKdfExpectsConcatenationKdfSpecAsKeySpec() throws Exception {
```

Comment:
should this be `counterKdfExpectsCounterKdfSpecAsKeySpec`? we're requesting a Counter KDF factory and asserting that it doesn't accept a PBEKeySpec